src/main/java/tigase/xmpp/impl/SaslAuthAbstract.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Tigase XMPP Server - The instant messaging server
+ * Copyright (C) 2004 Tigase, Inc. (office@tigase.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, version 3 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. Look for COPYING file in the top folder.
+ * If not, see http://www.gnu.org/licenses/.
+ */
+
+package tigase.xmpp.impl;
+
+import tigase.auth.*;
+import tigase.db.AuthRepository;
+import tigase.db.TigaseDBException;
+import tigase.kernel.beans.Inject;
+import tigase.server.Command;
+import tigase.server.Packet;
+import tigase.server.Priority;
+import tigase.xml.Element;
+import tigase.xmpp.NotAuthorizedException;
+import tigase.xmpp.StanzaType;
+import tigase.xmpp.XMPPProcessorIfc;
+import tigase.xmpp.XMPPResourceConnection;
+import tigase.xmpp.jid.BareJID;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+abstract public class SaslAuthAbstract
+		extends AbstractAuthPreprocessor
+		implements XMPPProcessorIfc {
+
+	protected final static String SASL_SERVER_KEY = "SASL_SERVER_KEY";
+	protected final static String ALLOWED_SASL_MECHANISMS_KEY = "allowed-sasl-mechanisms";
+
+	private static final Logger log = Logger.getLogger(SaslAuthAbstract.class.getName());
+
+	public enum ElementType {
+		ABORT,
+		AUTHENTICATE,
+		CHALLENGE,
+		FAILURE,
+		RESPONSE,
+		SUCCESS,
+		CONTINUE,
+		NEXT,
+		DATA,
+		UPGRADE,
+		PARAMETERS,
+		HASH;
+
+		private static final Map<String, ElementType> ALL_TYPES = Arrays.stream(ElementType.values())
+				.collect(Collectors.toMap(ElementType::getElementName, Function.identity()));
+		private final String elementName;
+
+		public static ElementType parse(String name) {
+			return ALL_TYPES.get(name);
+		}
+
+		ElementType() {
+			this.elementName = name().toLowerCase();
+		}
+
+		public String getElementName() {
+			return elementName;
+		}
+	}
+
+	protected final Map<String, Object> props = new HashMap<>();
+	@Inject
+	protected TigaseSaslProvider saslProvider;
+
+	@Override
+	public int concurrentQueuesNo() {
+		return super.concurrentQueuesNo() * 4;
+	}
+
+	protected Element createReply(final SaslAuth2.ElementType type, final String cdata) {
```

Comment:
As I'm not convinced on using "unified" ElementType, maybe we could just use `String` as a `type` here?

Reply:
`ElementType` should have been moved to SaslAbstract. As for the usage - I'm not convinced to use String here, if there is opposition then using separate enums for both Sasl and separate `createReply` would be better.

Still - this is a set of possible element types, which seems quite "share-able"

Reply:
It's currently stuck? And if/when it's published and then if/when it has any changes we can decouple the enum if that's required? Maintaining a copy simply because something could happen seems inefficient.

Reply:
Maintaining a copy of the list of names used by a particular XEP makes sense, as we can update enum exact implementation of the XEP in our code without impact on other parts of Tigase and also without making any API changes not connected to the XEP implementation that we are updating.

If you wish to leave it as it is, we can do that, but I'm not convinced that this is a way to go.

Reply:
Went with two separate enums.